### PR TITLE
caching bridge optimizer copy improvement

### PR DESCRIPTION
**Pull request by ndinsmore**
This pull creates a new `add_constraints` method for bridge optimizers than uses it `copyconstraints!`.  The end result is a significant model creation improvement for Clp.  Benchmarks to follow.   

The associated PRs are:
https://github.com/JuliaOpt/Clp.jl/pull/57
https://github.com/JuliaOpt/LinQuadOptInterface.jl/pull/96

The benchmarking results for Clp show a 40x improvement in the time it takes to copy large models from the cache to the solver:
https://github.com/JuliaOpt/Clp.jl/pull/57#issuecomment-471838224

### CODE REVIEW

File: src/Utilities/copy.jl [hunk 1]
```diff
@@ -151,14 +151,14 @@ Copy the constraints of type `F`-in-`S` from the model `src` the model `dest` an
 function copyconstraints!(dest::MOI.ModelLike, src::MOI.ModelLike, copy_names::Bool, idxmap::IndexMap, ::Type{F}, ::Type{S}) where {F<:MOI.AbstractFunction, S<:MOI.AbstractSet}
     # Copy constraints
     cis_src = MOI.get(src, MOI.ListOfConstraintIndices{F, S}())
-    for ci_src in cis_src
-        f_src = MOI.get(src, MOI.ConstraintFunction(), ci_src)
-        f_dest = mapvariables(idxmap, f_src)
-        s = MOI.get(src, MOI.ConstraintSet(), ci_src)
-        ci_dest = MOI.add_constraint(dest, f_dest, s)
+    f_src = MOI.get(src, MOI.ConstraintFunction(), cis_src)
+    # idxmap is wraped with [] to signify singlton
```
Comment (blegat) [inline src/Utilities/copy.jl]: Missing e in singleton

File: src/Bridges/bridgeoptimizer.jl [hunk 2]
```diff
@@ -323,6 +323,25 @@ function MOI.add_constraint(b::AbstractBridgeOptimizer, f::MOI.AbstractFunction,
         return MOI.add_constraint(b.model, f, s)
     end
 end
+function MOI.add_constraints(b::AbstractBridgeOptimizer, f::Vector{F},
+                            s::Vector{S}) where {F <: MOI.AbstractFunction, S <: MOI.AbstractSet}
```
Comment (blegat) [inline src/Bridges/bridgeoptimizer.jl]: Missing space for indent

File: src/Bridges/bridgeoptimizer.jl [hunk 3]
```diff
@@ -323,6 +323,25 @@ function MOI.add_constraint(b::AbstractBridgeOptimizer, f::MOI.AbstractFunction,
...
+    if is_bridged(b, F, S)
+        # We compute `BridgeType` first as `concrete_bridge_type` calls
+        # `bridge_type` which might throw an `UnsupportedConstraint` error in
+        # which case, we do not want any modification to have been done
+        BridgeType = concrete_bridge_type(b, F, S)
+        # `add_constraint` might throw an `UnsupportedConstraint` but no
+        # modification has been done in the previous line
+        cis = MOI.add_constraints(b.bridged, f, s)
+        for (n,ci) in enumerate(cis)
```
Comment (blegat) [inline src/Bridges/bridgeoptimizer.jl]: Missing space after comma

File: src/Utilities/copy.jl (hunk 1 above)
Comment (blegat) [reply src/Utilities/copy.jl]: Two p's in wrapped

File: src/Utilities/copy.jl [hunk 4]
```diff
@@ -151,14 +151,14 @@ Copy the constraints of type `F`-in-`S` from the model `src` the model `dest` an
...
+    f_dest = mapvariables.([idxmap], f_src)
```
Comment (blegat) [inline src/Utilities/copy.jl]: The common approach is `Ref (idxmap)`

File: src/Utilities/copy.jl [hunk 5]
```diff
@@ -151,14 +151,14 @@ Copy the constraints of type `F`-in-`S` from the model `src` the model `dest` an
...
+    s = MOI.get(src, MOI.ConstraintSet(), cis_src)
+    cis_dest = MOI.add_constraints(dest, f_dest, s)
+    for (ci_src,ci_dest) in zip(cis_src, cis_dest)
```
Comment (blegat) [inline src/Utilities/copy.jl]: Missing space after comma

File: src/Utilities/copy.jl [hunk 6]
```diff
@@ -184,8 +184,9 @@ function default_copy_to(dest::MOI.ModelLike, src::MOI.ModelLike, copy_names::Bo
 
     # Copy variables
     vis_src = MOI.get(src, MOI.ListOfVariableIndices())
-    for vi in vis_src
-        idxmap.varmap[vi] = MOI.add_variable(dest)
+    vars=MOI.add_variables(dest, length(vis_src))
```
Comment (blegat) [inline src/Utilities/copy.jl]: Missing spaces around =

File: src/Utilities/copy.jl [hunk 7]
```diff
@@ -184,8 +184,9 @@ function default_copy_to(dest::MOI.ModelLike, src::MOI.ModelLike, copy_names::Bo
...
+    for (vi,var) in zip(vis_src, vars)
```
Comment (blegat) [inline src/Utilities/copy.jl]: Missing space after comma

File: src/Bridges/bridgeoptimizer.jl [hunk 8]
```diff
@@ -323,6 +323,25 @@ function MOI.add_constraint(b::AbstractBridgeOptimizer, f::MOI.AbstractFunction,
...
     end
 end
+function MOI.add_constraints(b::AbstractBridgeOptimizer, f::Vector{F},
+                            s::Vector{S}) where {F <: MOI.AbstractFunction, S <: MOI.AbstractSet}
+    if is_bridged(b, F, S)
+        # We compute `BridgeType` first as `concrete_bridge_type` calls
+        # `bridge_type` which might throw an `UnsupportedConstraint` error in
+        # which case, we do not want any modification to have been done
+        BridgeType = concrete_bridge_type(b, F, S)
+        # `add_constraint` might throw an `UnsupportedConstraint` but no
```
Comment (blegat) [inline src/Bridges/bridgeoptimizer.jl]: Do we have a test to cover this ?

Comment (ndinsmore) [reply src/Bridges/bridgeoptimizer.jl]: To be honestly this was in the non-plural method, and I am not sure that I fully under how to check it.

Comment (blegat) [reply src/Bridges/bridgeoptimizer.jl]: Just write "See `add_constraint` for why we we call `concrete_bridge_type` separately before `add_constraint`" then

File: src/Bridges/bridgeoptimizer.jl [hunk 9]
```diff
@@ -330,9 +330,11 @@ function MOI.add_constraints(b::AbstractBridgeOptimizer, f::Vector{F},
         # We compute `BridgeType` first as `concrete_bridge_type` calls
         # `bridge_type` which might throw an `UnsupportedConstraint` error in
         # which case, we do not want any modification to have been done
+        # See add_constraint for why we we call `concrete_bridge_type` separately 
+        # before `add_constraints`
```
Comment (blegat) [inline src/Bridges/bridgeoptimizer.jl]: You can remove the rest of the comments now

File: src/Bridges/bridgeoptimizer.jl [hunk 10]
```diff
@@ -323,6 +323,25 @@ function MOI.add_constraint(b::AbstractBridgeOptimizer, f::MOI.AbstractFunction,
...
+                             s::Vector{S}) where { F <: MOI.AbstractFunction,
+                             S <: MOI.AbstractSet}
+    if is_bridged(b, F, S)
+        # See add_constraint for why we we call `concrete_bridge_type` separately
+        # before `add_constraints`
+        BridgeType = concrete_bridge_type(b, F, S)
+        # `add_constraints` might throw an `UnsupportedConstraint` but no
+        #  modification has been done in the previous line
+        cis = MOI.add_constraints(b.bridged, f, s)
+        for (n,ci) in enumerate(cis)
```
Comment (blegat) [inline src/Bridges/bridgeoptimizer.jl]: Missing space, `(n,ci)` -> `(n, ci)`

**Pull request merged**